feat: add support to load items using dat file from spr

MUN1Z opened the pull request:
# Description

PT-BR:
Foi adicionado o suporte para carregar os itens a partir de um .Dat de uma Spr e Dat.
Com isso podemos continuar utilizando editores de spr e dat como o object builder para agilizar o desenvolvimento de servidores derivados.

O assets editor quebra um galho, mas contem diversos problemas, ao tentar utilizar o mesmo algumas vezes para criar servidores derivados no canary, acabei desistindo por conta da grande quantidade de problemas que ele apresenta ao tentar trabalhar com ele por algumas horas aberto modificando centenas de itens por hora.

Com esse suporte o desenvolvimento de servidores derivados como Naruto, Pokemon, Dragon Ball e entre outros, se torna mais simples, e mais similar ao que toda a comunidade de derivados já está acostumado.

Quebrando uma barreia enorme entre a comunidade de derivados que se mantem na tfs por conta das facilidades dos editores.

Pelo menos até conseguirmos um editor para assets tão bom quanto o ObjectBuilder e a finalizacão do Remeres 100% funcional com suporte a assets.

EN:
Support for loading items from a .Dat of a Spr and Dat has been added.
With this we can continue using spr and dat editors like the object builder to speed up the development of derived servers.

The asset editor is a great solution, but it has several problems. When trying to use it a few times to create derivative servers in the canary, I ended up giving up due to the large number of problems it presents when trying to work with it for a few hours open, modifying hundreds of items per hour.

With this support, the development of derivative servers such as Naruto, Pokemon, Dragon Ball and others becomes simpler, and more similar to what the entire derivatives community is already accustomed to.

Breaking a huge barrier among the derivatives community that remains on tfs due to the editors' facilities.

At least until we get an asset editor as good as ObjectBuilder and the finalization of 100% functional Remeres with asset support.

https://github.com/user-attachments/assets/b93562d7-bc5e-4e41-b12a-9ff5eea4bf1e

https://github.com/user-attachments/assets/d282e1b1-1710-44a7-aee5-bd33d75044ba

https://github.com/user-attachments/assets/1c5ff96b-8972-45c3-9fee-9041e1abf330

https://github.com/user-attachments/assets/f1f27bf1-fcd8-4451-84cf-504b4e410334

  - [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested

Enable this
![image](https://github.com/user-attachments/assets/a3d9d285-b8d3-4dd8-a7f5-75ccf862cb76)

  - [x] enable in config.lua the flag "loadItemsFromSprDat"
  - [x] have a Tibia.dat (converted from 13.40 assets using SpiderClientConverter and checking all properties on .Dat export) into "data/items/Tibia.dat"

MUN1Z committed:
feat: add support to load items using dat file from spr.

MUN1Z committed:
Merge branch 'main' into feature/support_dat_items

MUN1Z committed:
Merge branch 'main' into feature/support_dat_items

MUN1Z committed:
Merge branch 'main' into feature/support_dat_items

MUN1Z committed:
Merge branch 'feature/support_dat_items' of https://github.com/tibia-devs/canary into feature/support_dat_items

murilo09 commented:
could you check the failed builds?

murilo09 commented on config.lua.dist:
```diff
@@ -588,3 +588,6 @@ metricsPrometheusAddress = "0.0.0.0:9464"
 metricsOstreamInterval = 1000
+
+-- NOTE: load items using dat file from old spr and dat cip files, or use default protobuff file if this false
+loadItemsFromSprDat = true
```
```suggestion
-- NOTE: load items using dat file from old spr and dat cip files, or use default protobuf file if false
loadItemsFromSprDat = false
```

murilo09 commented on src/items/items.cpp:
```diff
@@ -289,6 +294,320 @@ bool Items::loadFromXml() {
+					break;
+
+				case DatAttrMarket: {
+					auto category = props.read<uint16_t>();
```
unused

murilo09 commented on src/items/items.cpp:
```diff
@@ -289,6 +294,320 @@ bool Items::loadFromXml() {
+
+				case DatAttrMarket: {
+					auto category = props.read<uint16_t>();
+					auto tradeAsObjectId = props.read<uint16_t>();
```
unused

murilo09 commented on src/items/items.cpp:
```diff
@@ -289,6 +294,320 @@ bool Items::loadFromXml() {
+				case DatAttrMarket: {
+					auto category = props.read<uint16_t>();
+					auto tradeAsObjectId = props.read<uint16_t>();
+					auto showAsObjectId = props.read<uint16_t>();
```
unused

murilo09 commented on src/items/items.cpp:
```diff
@@ -289,6 +294,320 @@ bool Items::loadFromXml() {
+					auto tradeAsObjectId = props.read<uint16_t>();
+					auto showAsObjectId = props.read<uint16_t>();
+					auto name = props.readString();
+					auto vocation = props.read<uint16_t>();
```
unused

murilo09 commented on src/items/items.cpp:
```diff
@@ -289,6 +294,320 @@ bool Items::loadFromXml() {
+					auto showAsObjectId = props.read<uint16_t>();
+					auto name = props.readString();
+					auto vocation = props.read<uint16_t>();
+					auto minimumLevel = props.read<uint16_t>();
```
unused

dudantas commented:
You need to change here: https://github.com/opentibiabr/canary/blob/304656677dac994ce2784b9b5042f8c763806e93/src/canary_server.cpp#L348

For send the "Tibia.dat" message instead of "appearances.dat" depending from config.lua key

You can replace "appearances.dat" with "items.dat" everywhere in the code. This will simplify things since only one method will be used, making it unnecessary to have both names.

majestyotbr commented:
Closed due to inactivity.
Feel free to reopen if you intend to finish this pull request.

dudantas commented:
> Closed due to inactivity. Feel free to reopen if you intend to finish this pull request.

I'm going to finish it, the owner authorized me to make modifications to the pull request, so I reopened it to be able to make the necessary adjustments.

majestyotbr commented:
> > Closed due to inactivity. Feel free to reopen if you intend to finish this pull request.
> 
> I'm going to finish it, the owner authorized me to make modifications to the pull request, so I reopened it to be able to make the necessary adjustments.

Ok.

dudantas committed:
fix: working

dudantas committed:
Merge branch 'main' into pr/3159

dudantas committed:
fix: correctly reading the outfits and effects

Updating the items.dat (for more actually tibia version)
Added new flags
Reading new flags
Reading outfits/mounts and effects from dat

dudantas committed:
remove: unused struct

dudantas committed:
fix: default movable to true

items.dat doesn't have the "is movable" flag. Only the "not movable"

murilo09 commented on config.lua.dist:
```diff
@@ -605,3 +605,6 @@ metricsPrometheusAddress = "0.0.0.0:9464"
 metricsEnableOstream = false
 metricsOstreamInterval = 1000
+
+-- NOTE: load items using dat file from old spr and dat cip files, or use default protobuff file if this false
```
```suggestion
-- NOTE: load items using dat file from old spr and dat cip files, or use default protobuf file if false
```

murilo09 commented on src/items/items.cpp:
```diff
@@ -289,6 +294,356 @@ bool Items::loadFromXml() {
+				case DatAttrWriteableOnce:
+					item.canReadText = true;
+					item.maxTextLen = props.read<uint16_t>();
+					break;
```
```suggestion
				case DatAttrWriteable:
				case DatAttrWriteableOnce:
					item.canReadText = true;
					item.maxTextLen = props.read<uint16_t>();
					break;
```

murilo09 commented on src/items/items.cpp:
```diff
@@ -289,6 +294,356 @@ bool Items::loadFromXml() {
+
+		item.id = id;
+
+		uint8_t icount = 0, attr = -1;
```
```suggestion
		uint8_t icount = 0;
		uint8_t attr = -1;
```

dudantas committed:
Apply suggestions from code review

Co-authored-by: murilo09 <78226931+murilo09@users.noreply.github.com>

dudantas committed:
Merge branch 'main' into feature/support_dat_items

dudantas committed:
improve: sonar requests and others changes

dudantas commented:
I've made all the changes I thought were necessary. It's ready for test and review.

Now, have:
config.lua tag for set the "dat" filename
Reading the outfits/mounts and effects (magic and missiles) from the dat.
Others codes improvements.

dudantas committed:
add: missing tag

dudantas committed:
change: log to debug

dudantas committed:
fix: reduce enum verbose

carlosmiguelgit commented:
Qual possibilidade de fazer o .dat ler todas as flasgs atuais do appearences fazendo as coisas que dependem do protobuf funcionarem?

Zbizu commented:
@carlosmiguelgit 
seems like it currently works with one version only

@dudantas
I'll be adding my own implementation that handles multiple versions together with resourceId system in near future

carlosmiguelgit commented:
@dudantas may you share your own implementation for that please?

dudantas commented:
> @dudantas may you share your own implementation for that please?

I don't understand. What implementation? I have nothing on this. Read again what zbizu said; he said he will share his implementation.

carlosmiguelgit commented:
@Zbizu could you?
